### PR TITLE
8221360: Eliminate Shared_DirtyCardQ_lock

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to the handling of a very rare race condition in
concurrent refinement.  There is a situation where we many be unable to
process a card and need to re-dirty and re-enqueue it for later processing.

The old code handled this using a shared dirty card queue, which various
other places needed to know about as well.  This was the user of the
Shared_DirtyCardQ_lock.  It was the only remaining Mutex whose rank is based
on Mutex::access.

An option for eliminating that lock was to use a lock-free queue.  But
rather than take on that complexity, we instead brute-force the problem.
The new implementation simply allocates and enqueues an entire buffer for
the one card that needs to be reprocessed.  Those operations are already
lock-free.  The extra cost doesn't really matter, since this situation
almost never arises.

This is now done inline in refine_card_concurrently.  It's not worth the
overhead of a class for what is now just a few lines of code needed in
exactly one place.  Places that knew about the shared queue in order to
flush/reset it at certain points no longer have such responsibilities.

This change also renders JDK-8214997 moot, since there aren't any 'access'
locks.
https://bugs.openjdk.java.net/browse/JDK-8214997
Crash holding 'access' lock can deadlock in JVMCI compiler thread

This change also allows the removal of Mutex::access.  That's being left to
a new enhancement, since I think it also allows removing Mutex::tty,
reverting back to using Mutex::event for the tty_lock.  That is, essentially
undo JDK-8214315.
https://bugs.openjdk.java.net/browse/JDK-8272480

Testing:
mach5 tier1-5

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8221360](https://bugs.openjdk.java.net/browse/JDK-8221360): Eliminate Shared_DirtyCardQ_lock

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5118/head:pull/5118` \
`$ git checkout pull/5118`

Update a local copy of the PR: \
`$ git checkout pull/5118` \
`$ git pull https://git.openjdk.java.net/jdk pull/5118/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5118`

View PR using the GUI difftool: \
`$ git pr show -t 5118`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5118.diff">https://git.openjdk.java.net/jdk/pull/5118.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -1688,12 +1687,20 @@ void G1RemSet::refine_card_concurrently(CardValue* const card_ptr,
...
+  // the thread-local queue, because that might be the queue that is being
+  // processed by us; we could be a Java thread conscripted to perform
+  // refinement on our queue's current buffer.  This situation only arises
+  // from an extremely rare race condition in the memregion iteration, so
+  // it's not worth any significant development effort or clever lock-free
+  // queue implementation.  Instead we use brute force, allocating and
+  // enqueuing an entire buffer for just this card.
   *card_ptr = G1CardTable::dirty_card_val();
-  G1BarrierSet::shared_dirty_card_queue().enqueue(card_ptr);
+  G1DirtyCardQueueSet& dcqs = G1BarrierSet::dirty_card_queue_set();
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1699]: This is just a suggestion: maybe put this code to put and enqueue that card into an extra buffer into an extra method. It does not seem to fit just here.

**Review by tschatzl — APPROVED** [review]
Another reason probably why these cards are put into the shared DCQS was that that card will be processed at a time when that situation can't re-occur. So this change might cause that card (buffer) to be re-enqueued and reprocessed many times while waiting for that object to get fully visible.
One option to avoid this would be to temporarily raise the refinement threshold to avoid reprocessing, but it is very hard to know when it can be lowered (apart from complications actually doing that).
Another would be to use a non-dcqs lock-free container (e.g. a segmented array) to store all these entries (there may be some implementations around), but that's probably not worth it either, and somewhat ugly to transfer to actual dcqs later for processing.
I assume and I agree that this is a very rare occurrence, so we probably get away with this solution.

**Comment by kimbarrett** [comment]
> Another reason probably why these cards are put into the shared DCQS was that that card will be processed at a time when that situation can't re-occur. So this change might cause that card (buffer) to be re-enqueued and reprocessed many times while waiting for that object to get fully visible.

Dirty card buffers are processed in fifo order, so this single-card buffer
starts at the end of the line. Also, we generally keep some number of cards in
the queue. So it's going to take some time for this card to make it through
all that and be reconsidered.

**Comment by tschatzl** [comment]
> > Another reason probably why these cards are put into the shared DCQS was that that card will be processed at a time when that situation can't re-occur. So this change might cause that card (buffer) to be re-enqueued and reprocessed many times while waiting for that object to get fully visible.
> 
> Dirty card buffers are processed in fifo order, so this single-card buffer
> starts at the end of the line. Also, we generally keep some number of cards in
> the queue. So it's going to take some time for this card to make it through
> all that and be reconsidered.

I agree that there should typically be enough delay until that buffer/card gets reprocessed.

**Review by walulyai — APPROVED** [review]

**Comment by kimbarrett** [comment]
Thanks @tschatzl and @walulyai for reviews.

File: src/hotspot/share/gc/g1/g1RemSet.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:1706]: I split that out into new helper function enqueue_for_reprocessing.

**Review by tschatzl — APPROVED** [review]
Still good, thanks.

**Review by walulyai — APPROVED** [review]

**Pull request closed without merging**